rust/lance-table/src/rowids.rs

```diff
@@ -0,0 +1,266 @@
+    }
+}
+
+impl RowIdSequence {
+    pub fn iter(&self) -> impl DoubleEndedIterator<Item = u64> + '_ {
+        self.0.iter().flat_map(|segment| segment.iter())
+    }
+
+    pub fn len(&self) -> u64 {
+        self.iter().count() as u64
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Combines this row id sequence with another row id sequence.
+    pub fn extend(&mut self, other: Self) {
+        // If the last element of this sequence and the first element of next
+        // sequence are ranges, we might be able to combine them into a single
+        // range.
+        if let (Some(U64Segment::Range(range1)), Some(U64Segment::Range(range2))) =
+            (self.0.last(), other.0.first())
+        {
+            if range1.end == range2.start {
+                let new_range = U64Segment::Range(range1.start..range2.end);
+                self.0.pop();
+                self.0.push(new_range);
+                self.0.extend(other.0.into_iter().skip(1));
+                return;
+            }
+        }
+        self.0.extend(other.0);
+    }
+
+    /// Mark a set of row ids as deleted. Their value will be replaced with tombstones.
+    pub fn delete(&mut self, row_ids: impl IntoIterator<Item = u64>) {
+        // Order the row ids by position in which they appear in the sequence.
+        let (row_ids, offsets) = self.find_ids(row_ids);
+
+        let capacity = self.0.capacity();
+        let old_segments = std::mem::replace(&mut self.0, Vec::with_capacity(capacity));
+        let mut remaining_segments = old_segments.as_slice();
+
+        for (segment_idx, range) in offsets {
+            let segments_handled = old_segments.len() - remaining_segments.len();
+            let segments_to_add = segment_idx - segments_handled;
+            self.0
+                .extend_from_slice(&remaining_segments[..segments_to_add]);
+            remaining_segments = &remaining_segments[segments_to_add..];
+
+            let segment;
+            (segment, remaining_segments) = remaining_segments.split_first().unwrap();
+
+            let segment_ids = &row_ids[range];
+            self.0.push(segment.delete(segment_ids));
+        }
+
+        // Add the remaining segments.
+        self.0.extend_from_slice(remaining_segments);
+    }
+
+    /// Find the row ids in the sequence.
+    ///
+    /// Returns the row ids sorted by their appearane in the sequence.
+    /// Also returns the segment index and the range where that segment's
+    /// row id matches are found in the returned row id vector.
+    fn find_ids(
+        &self,
+        row_ids: impl IntoIterator<Item = u64>,
+    ) -> (Vec<u64>, Vec<(usize, Range<usize>)>) {
+        // Often, the row ids will already be provided in the order they appear.
+        // So the optimal way to search will be to cycle through rather than
+        // restarting the search from the beginning each time.
+        let mut segment_iter = self.0.iter().enumerate().cycle();
```

Comment:
Perhaps the thing you are missing is we are re-using the same `segment_iter` in the `row_ids.into_iter().for_each()` loop. This means each new search for a row id will pick up right after the last one we found. The idea is this is more efficient in the common case where the row ids we are searching for are in the same order they appear in the segment. Instead of restarting our search at the beginning of the segment for each row id, we can just keep going from where we left off. This makes the sorted case `O(max(n, m))` instead of `O(n * m)` (`n` being the length of segment and `m` being number of row ids we are searching for).

The reason we limit each search to `self.0.len()` is so that if we are passed a non-existant row id we don't loop forever.